### ISSUE TITLE
Cannot auto `rolling' initialized session

### ISSUE THREAD

**Issue by Luluno01** [balderdashy/sails#4254]
**Sails version**: 0.12.14
**Node version**: v9.2.0
**NPM version**: 3.10.10 (npm3)
**DB adapter name**: sails-mysql for model storage and connect-redis for session storage
**DB adapter version**: 0.11.5 (sails-mysql) and 3.3.2 (connect-redis)
**Operating system**: CentOS 7 x86-64

<hr/>

# What happened to my app?

Dear developers,

I was configuring the session of my app and found the auto `rolling` feature for sessions seems to not work, that is, the `session.maxAge` of the corresponding request **will not** be auto reseted if `session` is not modified.

I have read https://github.com/expressjs/session#cookie carefully but I still can not figure out what is wrong with my `config/local.js` and my app :( .

For some political reason I cannot access to http://bit.ly/sails-issue-repro so I am not sure whether my issue reproducing repo meets your requirements :( .

# My configurations

Session storage adapter: `connect-redis`

Redis version: `4.0.6`

`config/local.js`:

```
module.exports = {
  port: 80,
  session: {
    adapter: 'connect-redis',
    host: '127.0.0.1',
    ttl: 7 * 24 * 3600, // One week
    db: 0,
    prefix: 'sess:',
    cookie: {
      resave: false,
      maxAge: 7 * 24 * 3600 * 1000, // One week
      rolling: true,
      saveUninitialized: false
    },
    routesDisabled: ['GET /js/*', 'GET /css/*', 'GET /images/*']
  },
  models: {
    migrate: 'safe',
    connection: 'localDiskDb'
    // connection: 'mysql'
  }
}
```

# Issue reproducing repository

[sails-issue-repro-session-rolling](https://github.com/Luluno01/sails-issue-repro-session-rolling)

# Issue reproducing steps

1. Run app.
2. Visit http://localhost/test/wr?str=rua and check the expiration date of the session cookie on a browser. This request will modify `session` and therefore the `session` is initialized and will be saved.
3. Wait for some time, say, 1 minute.
4. Visit http://localhost/test/rd and check the expiration date of the session cookie on a browser again. This request **will not** modify `session`.

You will find the expiration date of the session cookie will not be reset until you modify `session` by visiting http://localhost/test/wr?str=someDifferentString again.

# Finally

The following configurations cannot help, too.

```
// config/local.js
module.exports = {
  port: 80,
  session: {
    adapter: 'connect-redis',
    host: '127.0.0.1',
    ttl: 7 * 24 * 3600, // One week
    db: 0,
    prefix: 'sess:',
    cookie: {
      resave: true, // Hey, there
      maxAge: 7 * 24 * 3600 * 1000, // One week
      rolling: true,
      saveUninitialized: false
    },
    routesDisabled: ['GET /js/*', 'GET /css/*', 'GET /images/*']
  },
  models: {
    migrate: 'safe',
    connection: 'localDiskDb'
    // connection: 'mysql'
  }
}
```

or 

```
// config/local.js
module.exports = {
  port: 80,
  session: {
    adapter: 'connect-redis',
    host: '127.0.0.1',
    ttl: 7 * 24 * 3600, // One week
    db: 0,
    prefix: 'sess:',
    cookie: {
      resave: true, // Hey, there
      maxAge: 7 * 24 * 3600 * 1000, // One week
      rolling: true,
      saveUninitialized: true // Hey, there
    },
    routesDisabled: ['GET /js/*', 'GET /css/*', 'GET /images/*']
  },
  models: {
    migrate: 'safe',
    connection: 'localDiskDb'
    // connection: 'mysql'
  }
}
```

That's all. I suspect this is a bug, probably. I hope some would help me with this issue very soon XD.

**Comment by sailsbot** [comment]
Hi @Luluno01!  It looks like you missed a step or two when you created your issue.  Please edit your comment (use the pencil icon at the top-right corner of the comment box) and fix the following:

* Verify "My issue title is concise, on-topic and polite ("jst.js being removed from layout.ejs on lift" is good; "templates dont work" or "why is sails dumb" are not so good)."

As soon as those items are rectified, post a new comment (e.g. &ldquo;Ok, fixed!&rdquo;) below and we'll take a look.  Thanks!

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by Luluno01** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
Sorry to be a hassle, but it looks like your issue is still missing some required info.  Please double-check your initial comment and try again.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by Luluno01** [comment]
Ok, fixed!

**Comment by sailsbot** [comment]
@Luluno01 Thanks for posting, we'll take a look as soon as possible.

---

For help with questions about Sails, [click here](http://sailsjs.com/support).  If you&rsquo;re interested in hiring @sailsbot and her minions in Austin, [click here](http://sailsjs.com/studio).

**Comment by sgress454** [comment]
HI @Luluno01, thanks for the detailed post and the repo.  We're trying to wrap up the 1.0 release so it may take a little bit to check this out.  In the meantime, seems like a workaround would be a policy or custom hook that runs before every route and sets a key in the session to a random value, since it sounds like you're saying that as long as the session is modified, the expiration will be reset?

**Comment by Luluno01** [comment]
@sgress454 Thanks for your reply! The solution you provide sounds like a "quick patch" that could not tackle the root problem but would work with me.

Yes, I will try this out to fix my app temporarily. Please let me know if you great developers have any further information about this issue in the future!

Looking forward to your 1.0 release version :) . I hope this issue would not be found in the brand new version XD.

**Comment by Luluno01** [comment]
@sgress454 Alright, here is my custom session fixing middleware:

```
sessionFix: (req, res, next) => {
  if(req.session) {
    let keys = Object.keys(req.session); // ['cookie', '_sessionFix', ...]
    let fixKeyIndex = keys.indexOf('_sessionFix');
    if(fixKeyIndex != -1) { // Key found
      keys.splice(fixKeyIndex, 1);
    }
    if(keys.length > 1) { // Initialized session
      req.session._sessionFix = Math.random(); // Force the `maxAge` property to be reseted
    }
  }
  next();
},
```

I have tested this, works fine for now :) .

**Comment by sailsbot** [comment]
@Luluno01,@sailsbot,@sgress454: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by Luluno01** [comment]
I am waiting for a patch or a new version of sails.js : )

Keep it alive, please.

**Comment by sailsbot** [comment]
@Luluno01,@sailsbot,@sgress454: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by Luluno01** [comment]
Still waiting for a patch XD.

**Comment by sailsbot** [comment]
@Luluno01,@sailsbot,@sgress454: Hello, I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message and simply close the issue if possible.  On the other hand, if you are still waiting on a patch, please post a comment to keep the thread alive (with any new information you can provide).



If no further activity occurs on this thread within the next 3 days, the issue will automatically be closed.

Thanks so much for your help!

**Comment by luislobo** [comment]
@sailsbot seems like there seems to be something to be done here yet

**Comment by Luluno01** [comment]
Still waiting ( ͡° ͜ʖ ͡°)✧

**Comment by mikermcneil** [comment]
@Luluno01 thanks for your patience!

So to be honest, I doubt I'm going to be able to justify the development hours to patch this on the v0.12 version of Sails unless we have that need come up on a project.

That said, if you notice that this is an issue on v1.0 of Sails, I'll look into it personally.

**Comment by Luluno01** [comment]
@mikermcneil Thank you for your reply! Actually I can't wait to try out the latest version of Sails but I am afraid of not being able to have enough time to check this out on v1.0 of Sails recently. So should I close this issue and open a new issue if I find the same issue on v1.0 of Sails, or just leave this issue open?

Thanks for your help once again!

**Comment by alejandrocibilich** [comment]
@Luluno01, I believe you have a problem in the configuration. The property `rolling`, as well as `resave` and `saveUninitialized`, should all be in `sails.config.session` and not in `sails.config.session.cookie`. 
At least that's how it works in Sails v1. Took me a time debugging `express-session` package to realize I had made the same mistake.
